rusoto/credential/src/environment.rs

```diff
@@ -86,47 +151,61 @@ impl ProvideAwsCredentials for EnvironmentProvider {
 
     fn credentials(&self) -> Self::Future {
         EnvironmentProviderFuture {
-            inner: result(credentials_from_environment())
+            inner: result(credentials_from_environment(self)),
         }
     }
 }
 
 /// Grabs the Credentials from the environment.
-fn credentials_from_environment() -> Result<AwsCredentials, CredentialsError> {
-    let env_key = match non_empty_env_var(AWS_ACCESS_KEY_ID) {
-        Some(val) => val,
-        None => return Err(CredentialsError::new(E_NO_ACCESS_KEY_ID)),
-    };
-    let env_secret = match non_empty_env_var(AWS_SECRET_ACCESS_KEY) {
-        Some(val) => val,
-        None => {
-            return Err(CredentialsError::new(E_NO_SECRET_ACCESS_KEY))
-        }
-    };
+fn credentials_from_environment(
+    provider: &EnvironmentVariableProvider,
+) -> Result<AwsCredentials, CredentialsError> {
+    let env_key = get_critical_variable(provider.access_key_id_var())?;
+    let env_secret = get_critical_variable(provider.secret_access_key_var())?;
     // Present when using temporary credentials, e.g. on Lambda with IAM roles
-    let token = non_empty_env_var(AWS_SESSION_TOKEN);
-    // Mimik botocore's behavior, see https://github.com/boto/botocore/pull/1187.
-    let expires_at = match non_empty_env_var(AWS_CREDENTIAL_EXPIRATION) {
-        Some(val) => match DateTime::<FixedOffset>::parse_from_rfc3339(&val).map(|dt| dt.with_timezone(&Utc)) {
-            Ok(e) => Some(e),
-            Err(e) => return Err(CredentialsError::new(format!("{} '{}': {}", E_INVALID_EXPIRATION, val, e)))
-        },
+    let token = non_empty_env_var(&provider.session_token_var());
+    // Mimic botocore's behavior, see https://github.com/boto/botocore/pull/1187.
+    let var_name = provider.credential_expiration_var();
+    let expires_at = match non_empty_env_var(&var_name) {
+        Some(val) => Some(DateTime::<FixedOffset>::parse_from_rfc3339(&val)
+            .map(|dt| dt.with_timezone(&Utc))
+            .map_err(|e| {
+                CredentialsError::new(format!(
+                    "Invalid {} in environment '{}': {}",
+                    var_name, val, e
+                ))
+            })?),
         _ => None,
     };
-    Ok(AwsCredentials::new(
-        env_key,
-        env_secret,
-        token,
-        expires_at,
-    ))
+    Ok(AwsCredentials::new(env_key, env_secret, token, expires_at))
+}
+
+/// Force an error if we do not see the particular variable name in the env.
+fn get_critical_variable(var_name: String) -> Result<String, CredentialsError> {
+    match non_empty_env_var(&var_name) {
```

Comment:
Instead of the match expression, [`Option::ok_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else) could work well here.

Reply:
Agreed, I can make that change as well.